docs/proposals/compute-resource-metrics-api.md

```diff
@@ -0,0 +1,148 @@
+server will be much smaller.
+
+## Data gathering
+
+We will use a push based system. Each kubelet will periodically - every 10s -
+POST its derived metrics to the API server. Then, any users of the metrics can
+register as watchers to receive the new metrics when they are available.
+
+Users of the metrics may also periodically poll the API server instead of
+register as a watcher, having in mind that new data may only be available
+every 10 seconds. If any user requires metrics that are either more specific
+(e.g. last 1s) or updated more often, they should use the metrics pipeline via
+Heapster.
+
+The API server will not hold any of this data directly. For our initial
+purposes, it will hold the most recent metrics obtained from each node in
+etcd. Then, when polled for metrics, the API server would only serve said most
+recent data per node.
+
+Benchmarks will be run with etcd to see if it can keep up with the frequent
+writes of data. If it turns out that etcd doesn't scale well enough, we will
+have to switch to a different storage system.
+
+If a pod gets deleted, the API server will get rid of any metrics it may
+currently be holding for it.
+
+The clients watching the metrics data may cache it for longer periods of time.
+The clearest example would be Heapster.
+
+In the future, we might want to store the metrics differently:
+
+* via heapster - Since heapster keeps data for a period of time, we could
+  redirect requests to the API server to heapster instead of using etcd. This
+  would also allow serving metrics other than the latest ones.
+
+An edge case that this proposal doesn't take into account is kubelets being
+restarted. If any of them are, with a simple implementation they would lose
+historical data and thus take hours to gather enough information to provide
+relevant metrics again. We might want to use persistent storage directly or in
+the future to improve that situation.
+
+More information on kubelet checkpoints can be read on
+[#489](https://github.com/GoogleCloudPlatform/kubernetes/issues/489).
+
+## Data structure
+
+```Go
+type DerivedPodMetrics struct {
+	TypeMeta
+	ObjectMeta // should have pod name
+	PodMetrics ResourcesMetrics
+	// the key is the container name
+	Containers []struct {
+		ContainerReference *Container
+		Metrics            ResourcesMetrics
+	}
+}
+
+type DerivedNodeMetrics struct {
+	TypeMeta
+	ObjectMeta // should have node name
+	NodeMetrics      ResourcesMetrics
+	SystemContainers []struct {
+		ContainerReference *Container
+		Metrics            ResourcesMetrics
+	}
+}
+
+type ResourcesMetrics map[resource.Type]DerivedMetrics
+
+type DerivedMetrics struct {
+	// End time of all the time windows in Metrics
+	Time util.Time `json:"time"`
+
+	// Last overlapping 10s, 1m, 1h and 1d as a start
+	// Updated every 10s, so the 10s window is sequential and the rest are
+	// rolling.
+	Metrics map[time.Duration]Metrics
+}
+
+type Metrics struct {
```

Comment:
Since we are already calculating percentiles should we group it in? It's often indicative than mean. You could calculate 50th, 95th, 99th, 99.9th percentile in a single pass.